ocaml/xapi/xapi_clustering.ml

```diff
@@ -146,6 +146,65 @@ let get_required_cluster_stacks ~__context ~sr_sm_type =
   (* We assume that we only have one SM for each SR type, so this is only to satisfy type checking *)
   |> List.flatten
 
+let get_corosync_version () =
+  (* example output corosync -v
+     Corosync Cluster Engine, version '2.4.6'
+     Copyright (c) 2006-2009 Red Hat, Inc.
+  *)
+  try
+    let out, err =
+      Forkhelpers.execute_command_get_output "/usr/sbin/corosync" ["-v"]
+    in
+    debug "output of corosync -v: stdout: %s stderr: %s" out err ;
+    match Astring.String.cut ~sep:"version " out with
+    | Some (_left, version) -> (
+      match Astring.String.with_range ~first:1 ~len:1 version with
+      | "2" ->
+          debug "Currently using corosync2" ;
+          Cluster_stack.Corosync2
+      | "3" ->
+          debug "Currently using corosync3" ;
+          Cluster_stack.Corosync3
+      | mv ->
+          handle_error
+            (InternalError
+               (Printf.sprintf "invalid corosync version number %s" mv)
+            )
+    )
+    | None ->
+        handle_error (InternalError "cannot parse corosync -v")
+  with Forkhelpers.Spawn_internal_error (out, err, _status) ->
+    Unix_error
+      (Printf.sprintf "failed to switch corosync version out: %s, err :%s" out
+         err
+      )
+    |> handle_error
+
+let maybe_switch_cluster_stack_version ~__context ~cluster_stack_version =
+  match Context.get_test_clusterd_rpc __context with
+  | Some _ ->
+      debug "in unit test, not switching cluster stack verion"
+  | None -> (
+      if cluster_stack_version = 3L then
+        match get_corosync_version () with
```

Comment:
I think switching the corosync version and getting the current corosync version is a job for the clustering daemon. Can we add functions to the cluster interface for this?